[script] support using hostname as node_ip_address

SongGuyang opened the pull request:
## Why are these changes needed?
In our case, we need to deploy a ray cluster between two LANs and the node ips are not public. We need to use http proxy to find routes by hostname.
In current PR, we remove the logic of converting address to ip and only reserve the logic of converting local host to a reachable across nodes IP. After this change:
If we start ray node in this way:
`ray start --head  --node-ip-address={myhostname}`
We will use hostname as node_ip_address to communicate with each other between nodes.
But notice that we also should add `_node_ip_address="{myhostname}"` in driver side, otherwise the driver couldn't connect the raylet successfully in some cases.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

SongGuyang commented:
I also need to add a test case.

MissiontoMars commented on python/ray/_private/services.py:
```diff
@@ -446,15 +446,18 @@ def node_ip_address_from_perspective(address):
 
 
-def get_node_ip_address(address="8.8.8.8:53"):
+def get_node_ip_address(address="8.8.8.8:53", use_hostname=False):
```
```suggestion
def get_node_ip_address_or_hostname(address="8.8.8.8:53", use_hostname=False):
```

kfstorm replied:
I think `get_node_host` is enough.

MissiontoMars commented on src/ray/core_worker/core_worker.cc:
```diff
@@ -80,7 +80,8 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
       task_execution_service_work_(task_execution_service_) {
-  RAY_LOG(DEBUG) << "Constructing CoreWorker, worker_id: " << worker_id;
+  RAY_LOG(DEBUG) << "Constructing CoreWorker, worker_id: " << worker_id
+                 << ", node_ip_address: " << options_.node_ip_address;
```
Is the `options_.node_ip_address` may be a hostname? If so, we should refactor the variable name.

SongGuyang replied:
So many places are using this name. I'm not sure whether to rename. @raulchen What do you think?

SongGuyang replied:
Dashboard frontend also distinguishes hostname and ip. We also can not rename simply.

ericl reviewed (commented):
Is there a reason we can't make this the default, instead of adding a new option?

SongGuyang commented:
@ericl Do you mean use hostname by default ? I'm not sure that if there are some issues in some special scenes. I found a PR of @robertnishihara in 2017. https://github.com/ray-project/ray/pull/1177

SongGuyang commented:
@ericl @raulchen Can you take a look at this PR? I have rewritten this feature and the flag `--use-hostname` has been removed. I have updated the description.

ericl reviewed (commented):
If the change is indeed this simple, it seems ok, if esoteric. My two concerns here:
- This is a very niche feature. I don't think it's likely any other Ray deployment in the world would use this flag.
- It will be very easy for this to break without good unit / integration tests.

Have we explored alternatives to deploying Ray in this scenario? Even if it introduces more complexity in some component external to Ray (i.e., using a VPN), it may be more maintainable in the long run.

SongGuyang commented:
@ericl Actually, we have more than one scene which needs to use this feature. The main requirement scene is federated learning which is mentioned by @raulchen in the scheduled meeting https://docs.google.com/document/d/156k8y8TsIQeaOczpHMHKWUxGJfDyfAblubBKigJ06lg/edit#heading=h.gw1orpbtn9c9. In the federated learning network, they can't use VPN because of some secure issues. And we will use the `http_proxy` env to forward the grpc across a gateway. Another scene is that some users want to build a ray cluster in aliyun. Their ray nodes will also interact with each other across gateways.

> * It will be very easy for this to break without good unit / integration tests.
Yep, I'm thinking about how to add the unit tests. But I don't know how to emulate a two nodes cluster with different available hostnames. Do you have any suggestion?

ericl commented:
You could setup a local http proxy routing based on fake hostnames for testing. Not sure how easy this is in python.

SongGuyang commented:
> You could setup a local http proxy routing based on fake hostnames for testing. Not sure how easy this is in python.

I think it is not easy. Can I just add a single node test which use hostname to start head node and driver? It can guarantee the hostname feature works. And about the cases using ip, I think current tests is enough. 
If I add this test, can we merge this PR? Or we need to do more discuss about other alternatives?

SongGuyang commented:
And the precondition of the new test is that I can get a available hostname in the CI machine.

SongGuyang commented:
@ericl @raulchen I have added a test https://github.com/ray-project/ray/pull/20720/files#diff-6b3376d6d736a3f776cbda99b7038353a232a183f0979e8f04697d11ab05dd2eR123. Can we review and merge this?

ericl commented:
@mwtian could you review this for merge?

raulchen commented on python/ray/tests/test_basic_5.py:
```diff
@@ -119,5 +120,24 @@ def test_internal_kv(ray_start_regular):
+@pytest.mark.skipif(
+    os.environ.get("CI") and sys.platform != "linux",
+    reason="This test is only run on linux CI machines.")
+def test_using_hostname(ray_start_cluster):
```
```suggestion
def test_ray_init_using_hostname(ray_start_cluster):
```
Maybe move this to `ray_init_test`?

raulchen commented:
Overall I think it makes sense to directly use whatever users inputs are (hostname or IP). If tests pass, it should be go.

mwtian commented on python/ray/scripts/scripts.py:
```diff
@@ -654,11 +656,14 @@ def start(node_ip_address, address, port, redis_password, redis_shard_ports,
                 cli_logger.print("{} ray", cf.magenta("import"))
                 cli_logger.print(
-                    "ray{}init(address{}{}{})", cf.magenta("."),
+                    "ray{}init(address{}{}{}{})", cf.magenta("."),
```
In the case of joining an existing cluster (`address="auto"`), it might be possible to make the `_node_ip_address` parameter unnecessary, by reading the value from a running Raylet with `_find_address_from_flag()`. Ok to just leave a note or todo.

SongGuyang replied:
Good suggestion. Add a note here https://github.com/ray-project/ray/pull/20720/files#diff-15b2fb983deb109a406d72dff4c44e7d063a8274c6bfedaea2f29697066511ccR659.

mwtian commented on python/ray/_private/services.py:
```diff
@@ -447,7 +447,7 @@ def canonicalize_bootstrap_address(addr: str):
         addr = get_ray_address_from_environment()
     try:
-        bootstrap_address = address_to_ip(addr)
+        bootstrap_address = convert_local_host_if_needed(addr)
```
Can you update `ip:port` to `host:port` in the function comment?

SongGuyang replied:
Sure.

kfstorm commented on python/ray/tests/test_basic_5.py:
```diff
@@ -119,5 +120,24 @@ def test_internal_kv(ray_start_regular):
+        })
+
+    # Use `ray.init` to test the connection.
+    ray.init(address=cluster.address, _node_ip_address=hostname)
```
Let me make it clear. If we use IP addresses everywhere, we don't need to provide the `_node_ip_address` argument when starting a driver script, but it's needed if we use hostnames everywhere, right? Is there any way to omit `_node_ip_address` when hostnames are used?

SongGuyang replied:
Actually, in my one node test, the `_node_ip_address` can be omitted.

kfstorm replied:
Then maybe we should remove `_node_ip_address` from the test case?

SongGuyang replied:
I don't think we should remove `_node_ip_address`.  I think there are some special optimization for the case in which address is found automatically. Maybe it couldn't work in a multiple nodes cluster(I just guess). And here we must ensure  `_node_ip_address=hostname` could work.

SongGuyang commented on python/ray/tests/test_basic_5.py:
```diff
@@ -119,5 +120,24 @@ def test_internal_kv(ray_start_regular):
+
+    # Use `ray.init` to test the connection.
+    ray.init(address=cluster.address, _node_ip_address=hostname)
+
```
Good~

SongGuyang commented on python/ray/tests/test_basic_5.py:
```diff
@@ -119,5 +120,24 @@ def test_internal_kv(ray_start_regular):
+    ray.init(address=cluster.address, _node_ip_address=hostname)
+
+    # Clean up.
+    ray.shutdown()
```
OK, remove this.
